mqtt/mqtt-broker/src/translation.rs

```diff
@@ -0,0 +1,81 @@
+use crate::ClientEvent;
+use lazy_static::lazy_static;
+use mqtt3::proto;
+use regex::Regex;
+
+pub fn translate_incoming(client_id: &str, event: ClientEvent) -> ClientEvent {
+    match event {
+        ClientEvent::Subscribe(s) => ClientEvent::Subscribe(subscribe(client_id, s)),
+        ClientEvent::Unsubscribe(u) => ClientEvent::Unsubscribe(unsubscribe(client_id, u)),
+        ClientEvent::PublishFrom(p) => ClientEvent::PublishFrom(publish_from(client_id, p)),
+        e => e,
+    }
+}
+
+pub fn translate_outgoing(packet: proto::Packet) -> proto::Packet {
+    match packet {
+        proto::Packet::Publish(p) => proto::Packet::Publish(publish_to(p)),
+        p => p,
+    }
+}
+
+fn subscribe(client_id: &str, subscribe: proto::Subscribe) -> proto::Subscribe {
+    let proto::Subscribe {
+        packet_identifier,
+        mut subscribe_to,
+    } = subscribe;
+
+    for mut sub_to in &mut subscribe_to {
+        sub_to.topic_filter = to_new_topic(client_id, sub_to.topic_filter.clone());
+    }
+
+    proto::Subscribe {
+        packet_identifier,
+        subscribe_to,
+    }
+}
+
+fn unsubscribe(client_id: &str, mut unsubscribe: proto::Unsubscribe) -> proto::Unsubscribe {
+    unsubscribe.unsubscribe_from = unsubscribe
+        .unsubscribe_from
+        .into_iter()
+        .map(|t| to_new_topic(client_id, t))
+        .collect();
```

Comment:
@dmolokanov Is there a function instead of collect that will re-use the existing vector? I remember seeing something like that before

Reply:
Collect usually operates on iterators, not on vectors, and has corresponding semantic. If you want to reuse vector you need to iterate it with index and update in place. You may win some cycles, but it is subject to benchmark IMO